File: dart/bob/lib/bob.dart
class Bob {
  String response(String input) {
    if (input.isBlank) return "Fine. Be that way!";
    if (input.isAQuestionShouted) return "Calm down, I know what I'm doing!";
    if (input.isAQuestion) return "Sure.";
    if (input.isShouted) return "Whoa, chill out!";
    return "Whatever.";
  }
}

extension SentenceUtils on String {
  bool get isAQuestionShouted => isShouted && isAQuestion;

  bool get isAQuestion => trim().endsWith("?");

  bool get isBlank => trim().isEmpty;

  bool get isShouted => this == toUpperCase() && this != toLowerCase();
}

File: kotlin/bob/src/main/kotlin/Bob.kt
private val String.isAQuestion: Boolean
    inline get() = trimEnd().endsWith("?")

private val String.isShouted: Boolean
    inline get() = this == toUpperCase() && any { it.isLetter() }

object Bob {
    fun hey(sentence: String): String =
        when {
            sentence.isShouted && sentence.isAQuestion -> "Calm down, I know what I'm doing!"
            sentence.isShouted -> "Whoa, chill out!"
            sentence.isAQuestion -> "Sure."
            sentence.isBlank() -> "Fine. Be that way!"
            else -> "Whatever."
        }
}
